File: src/sdk/v2/python/examples/validator_operations.py
import asyncio
from vernachain_sdk import VernachainClient

async def main():
    async with VernachainClient(
        node_url="http://localhost:8545",
        api_key="your-api-key"
    ) as client:
        try:
            # Get validator set for shard 0
            validators = await client.get_validator_set(shard_id=0)
            print(f"Found {len(validators)} validators in shard 0")
            
            # Print validator details
            for validator in validators:
                print(f"\nValidator: {validator.address}")
                print(f"Stake: {validator.stake} VRN")
                print(f"Reputation: {validator.reputation}")
                print(f"Active: {validator.is_active}")
                print(f"Commission Rate: {validator.commission_rate or 0.0}%")
                
                if validator.delegators:
                    print(f"Delegators: {len(validator.delegators)}")
                    total_delegated = sum(d.get('amount', 0) for d in validator.delegators)
                    print(f"Total Delegated: {total_delegated} VRN")
            
            # Find the validator with the highest reputation
            best_validator = max(validators, key=lambda v: v.reputation)
            print(f"\nBest validator by reputation: {best_validator.address}")
            
            # Stake tokens on the best validator
            stake_amount = 100.0  # VRN tokens
            stake_result = await client.stake(
                amount=stake_amount,
                validator_address=best_validator.address
            )
            print(f"\nStaked {stake_amount} VRN on validator {best_validator.address}")
            print(f"Transaction hash: {stake_result.get('transaction_hash')}")
            
            # Monitor validator performance
            print("\nMonitoring validator performance...")
            for _ in range(3):  # Monitor for 3 intervals
                await asyncio.sleep(10)  # Wait 10 seconds between checks
                
                # Get updated validator info
                validators = await client.get_validator_set(shard_id=0)
                updated_validator = next(
                    v for v in validators 
                    if v.address == best_validator.address
                )
                
                print(f"\nValidator {updated_validator.address} status:")
                print(f"Total Blocks Validated: {updated_validator.total_blocks_validated}")
                print(f"Current Reputation: {updated_validator.reputation}")
                print(f"Total Stake: {updated_validator.stake} VRN")

        except Exception as e:
            print(f"Error: {e}")

if __name__ == "__main__":
    asyncio.run(main()) 

File: src/sdk/v2/javascript/examples/validator-operations.ts
import { VernachainClient } from '../src';

async function main() {
    const client = new VernachainClient({
        nodeUrl: 'http://localhost:8545',
        apiKey: 'your-api-key'
    });

    try {
        // Get validator set for shard 0
        const validators = await client.getValidatorSet(0);
        console.log(`Found ${validators.length} validators in shard 0`);

        // Print validator details
        validators.forEach(validator => {
            console.log(`\nValidator: ${validator.address}`);
            console.log(`Stake: ${validator.stake} VRN`);
            console.log(`Reputation: ${validator.reputation}`);
            console.log(`Active: ${validator.is_active}`);
            console.log(`Commission Rate: ${validator.commission_rate ?? 0.0}%`);
            console.log(`Total Blocks Validated: ${validator.total_blocks_validated}`);

            if (validator.delegators) {
                console.log(`Delegators: ${validator.delegators.length}`);
                const totalDelegated = validator.delegators.reduce(
                    (sum: number, d: { amount?: number }) => sum + (d.amount ?? 0),
                    0
                );
                console.log(`Total Delegated: ${totalDelegated} VRN`);
            }
        });

        // Find the validator with highest reputation
        const bestValidator = validators.reduce(
            (best, current) => current.reputation > best.reputation ? current : best,
            validators[0]
        );
        console.log(`\nBest validator by reputation: ${bestValidator.address}`);

        // Stake tokens on the best validator
        const stakeAmount = 100.0; // VRN tokens
        interface StakeResult { transactionHash: string }
        const stakeResult = await client.stake(stakeAmount, bestValidator.address) as StakeResult;
        console.log(`\nStaked ${stakeAmount} VRN on validator ${bestValidator.address}`);
        console.log(`Transaction hash: ${stakeResult.transactionHash}`);

        // Monitor validator performance
        console.log('\nMonitoring validator performance...');
        for (let i = 0; i < 3; i++) {
            await new Promise(resolve => setTimeout(resolve, 10000)); // Wait 10 seconds

            const updatedValidators = await client.getValidatorSet(0);
            const updatedValidator = updatedValidators.find(
                v => v.address === bestValidator.address
            );

            if (updatedValidator) {
                console.log(`\nValidator ${updatedValidator.address} status:`);
                console.log(`Total Blocks Validated: ${updatedValidator.total_blocks_validated}`);
                console.log(`Current Reputation: ${updatedValidator.reputation}`);
                console.log(`Total Stake: ${updatedValidator.stake} VRN`);
            }
        }
    } catch (error) {
        console.error('Error:', error);
    }
}

main();
